crates/goose/src/providers/openai_utils.rs

```diff
@@ -0,0 +1,560 @@
+                            .unwrap()
+                            .entry("tool_calls")
+                            .or_insert(json!([]));
+
+                        tool_calls.as_array_mut().unwrap().push(json!({
+                            "id": request.id,
+                            "type": "function",
+                            "function": {
+                                "name": sanitized_name,
+                                "arguments": tool_call.arguments.to_string(),
+                            }
+                        }));
+                    }
+                    Err(e) => {
+                        output.push(json!({
+                            "role": "tool",
+                            "content": format!("Error: {}", e),
+                            "tool_call_id": request.id
+                        }));
+                    }
+                },
+                MessageContent::ToolResponse(response) => {
+                    match &response.tool_result {
+                        Ok(contents) => {
+                            // Send only contents with no audience or with Assistant in the audience
+                            let abridged: Vec<_> = contents
+                                .iter()
+                                .filter(|content| {
+                                    content
+                                        .audience()
+                                        .is_none_or(|audience| audience.contains(&Role::Assistant))
+                                })
+                                .map(|content| content.unannotated())
+                                .collect();
+
+                            // Process all content, replacing images with placeholder text
+                            let mut tool_content = Vec::new();
+                            let mut image_messages = Vec::new();
+
+                            for content in abridged {
+                                match content {
+                                    Content::Image(image) => {
+                                        // Add placeholder text in the tool response
+                                        tool_content.push(Content::text("This tool result included an image that is uploaded in the next message."));
+
+                                        // Create a separate image message
+                                        image_messages.push(json!({
+                                            "role": "user",
+                                            "content": [convert_image(&image, image_format)]
+                                        }));
+                                    }
+                                    _ => {
+                                        tool_content.push(content);
+                                    }
+                                }
+                            }
+                            let concatenated_content = tool_content
```

Comment:
Converted content object list in message (ToolResponse message) to string to match api schema.   In [openai](https://platform.openai.com/docs/api-reference/chat/create), it accepted "string" or "array", but in groq it only accepts "string" based on their [api](https://console.groq.com/docs/api-reference#chat-create)

Reply:
@baxen is this related to the error you were seeing on the llama3 databricks endpoint?

Reply:
I've changed to only apply the concatenate logic to ollama and groq